### PR TITLE
GODRIVER-2989 A Dockerfile for local development

### PR DESCRIPTION
**Pull request by blink1073**
[GODRIVER-2989](https://jira.mongodb.org/browse/GODRIVER-2989)

## Summary
Adds a Dockerfile and a runner script.

## Background & Motivation
Make it easier to test in a reproducible Linux environment.

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -111,6 +111,19 @@ The usage of host.docker.internal comes from the [Docker networking documentatio
...
+Ensure ``docker`` is installed and running, and then run:
+
+```bash
+bash etc/run_docker.sh
+```
+
+The script takes an optional argument for the ``MAKEFILE_TARGET`` and allows for some environment variable overrides.
+The docker container will have the required binaries, including libmongocrypt.
+The entry script starts a mongodb topology, and then executes the desired ``MAKEFILE_TARGET``.
+
```
Comment (matthewdale) [inline docs/CONTRIBUTING.md:126]: Optional: Consider adding an example of how to override env vars.

> For example, to use the Makefile at `~/Makefile` and test against a sharded cluster, run:
> ```bash
> TOPOLOGY=sharded_cluster bash etc/run_docker.sh "~/Makefile"
> ```

**Review by matthewdale — APPROVED** [review]
Looks good 👍

File: etc/docker_entry.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+export PROJECT_ORCHESTRATION_HOME=$DRIVERS_TOOLS/.evergreen/orchestration
+export MONGO_ORCHESTRATION_HOME=$HOME
+
+# Clone DRIVERS_TOOLS if necessary.
+if [ ! -d $DRIVERS_TOOLS ]; then
+    git clone https://github.com/mongodb-labs/drivers-evergreen-tools.git $DRIVERS_TOOLS
+fi
+
+# Disable ipv6.
+sed -i "s/\"ipv6\": true,/\"ipv6\": false,/g" $PROJECT_ORCHESTRATION_HOME/configs/${TOPOLOGY}s/$ORCHESTRATION_FILE
```
[the remark this answers is not in the document]
Comment (blink1073) [reply etc/docker_entry.sh:18]: The connection is refused, this is a docker limitation.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  build-essential \
+  golang-go \
+  pkg-config \
+  libssl-dev \
+  make \
+  lsof \
+  net-tools \
+ && rm -rf /var/lib/apt/lists/*
+
+RUN export LIBMONGOCRYPT_TAG="1.8.0-alpha1" && \
```
[the remark this answers is not in the document]
Comment (blink1073) [reply Dockerfile:21]: Fixed

File: etc/docker_entry.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply etc/docker_entry.sh:18]: Added a note.

**Review by matthewdale — APPROVED** [review]
Looks great 👍 

Is there any chance we can switch the CI to use this as well, which would force us to keep this up-to-date?

**Comment by blink1073** [comment]
Well, the tests don't fully pass.  For me locally even with `bash etc/run_docker.sh test-short` I see the following failure:

```
go test  -tags= -timeout 60s -short ./...
ok  	go.mongodb.org/mongo-driver/benchmark	0.020s [no tests to run]
ok  	go.mongodb.org/mongo-driver/bson	0.212s
--- FAIL: TestTimeCodec (0.00s)
    --- FAIL: TestTimeCodec/UseLocalTimeZone (0.00s)
        --- FAIL: TestTimeCodec/UseLocalTimeZone/true (0.00s)
            time_codec_test.go:44:
                	Error Trace:	/src/bson/bsoncodec/time_codec_test.go:44
                	Error:      	Not equal:
                	            	expected: true
                	            	actual  : false
                	Test:       	TestTimeCodec/UseLocalTimeZone/true
                	Messages:   	Expected UTC: false, got UTC
FAIL
FAIL	go.mongodb.org/mongo-driver/bson/bsoncodec	0.040s
```

**Comment by blink1073** [comment]
I got the tests running, we we missing the `tzdata` package.  I also factored out a base image and put it in https://github.com/mongodb-labs/drivers-evergreen-tools/pull/352.

**Comment by blink1073** [comment]
Putting in draft until the upstream PR is merged.

**Comment by blink1073** [comment]
Okay, this is ready for review again.  I added a test, which meant adding a few test skips for docker-specific considerations.  I also created a script to install libmongocrypt to avoid repeating code and get us closer to the goal of less code in `config.yaml`.  This will also close [GODRIVER-2994](https://jira.mongodb.org/browse/GODRIVER-2994).

**Review by matthewdale — APPROVED** [review]
Awesome changes, looks good! 👍

**Pull request merged**